src/BinSkim.Rules/SignSecurely.cs

```diff
@@ -0,0 +1,456 @@
+            var certChain = (Native.CERT_SIMPLE_CHAIN)Marshal.PtrToStructure(simpleChains[0], typeof(Native.CERT_SIMPLE_CHAIN));
+
+            IntPtr[] chainElements = new IntPtr[certChain.cElement];
+            Marshal.Copy(certChain.rgpElement, chainElements, 0, chainElements.Length);
+            var certElement = (Native.CERT_CHAIN_ELEMENT)Marshal.PtrToStructure(chainElements[0], typeof(Native.CERT_CHAIN_ELEMENT));
+
+            var certContext = (Native.CERT_CONTEXT)Marshal.PtrToStructure(certElement.pCertContext, typeof(Native.CERT_CONTEXT));
+            certInfo = (Native.CERT_INFO)Marshal.PtrToStructure(certContext.pCertInfo, typeof(Native.CERT_INFO));
+
+            return ValidationResult.ERROR_SUCCESS;
+        }
+
+        private Native.WINTRUST_DATA InitializeWinTrustDataStruct(string filePath)
+        {
+            // See https://msdn.microsoft.com/en-us/library/windows/desktop/aa382384(v=vs.85).aspx
+            // which was used to drive data initialization, API use and comments in this code.
+
+            var winTrustData = new Native.WINTRUST_DATA();
+
+            var fileInfo = new Native.WINTRUST_FILE_INFO();
+            fileInfo.pcwszFilePath = filePath;   // Path to the signed file
+            fileInfo.cbStruct = (uint)Marshal.SizeOf(typeof(Native.WINTRUST_FILE_INFO));
+
+            winTrustData.cbStruct = (uint)Marshal.SizeOf(typeof(Native.WINTRUST_DATA));
+
+            winTrustData.pPolicyCallbackData = IntPtr.Zero;                                           // Use default code signing EKU
+            winTrustData.pSIPClientData = IntPtr.Zero;                                                // No data to pass to SIP
+            winTrustData.UIChoice = Native.UIChoice.WTD_UI_NONE;                                      // Disable all UI on execution
+            winTrustData.UIContext = 0;
+            winTrustData.UIContext = Native.UIContext.WTD_UICONTEXT_EXECUTE;                          // File is intended to be executed
+            winTrustData.UnionChoice = Native.UnionChoice.WTD_CHOICE_FILE;                            // We're verifying a file
+            winTrustData.RevocationChecks = Native.RevocationChecks.WTD_REVOKE_WHOLECHAIN;            // Revocation checking on whole chain.
+            winTrustData.dwProvFlags = Native.ProviderFlags.WTD_REVOCATION_CHECK_CHAIN_EXCLUDE_ROOT;  // Don't reach across the network
+            winTrustData.dwProvFlags |= Native.ProviderFlags.WTD_CACHE_ONLY_URL_RETRIEVAL;
+            winTrustData.pwszURLReference = null;                                                     // Reserved for future use
+
+            winTrustData.StateAction = Native.StateAction.WTD_STATEACTION_VERIFY;
+            winTrustData.hWVTStateData = IntPtr.Zero; // This value set by API call
+
+            winTrustData.File = Marshal.AllocHGlobal(Marshal.SizeOf(typeof(Native.WINTRUST_FILE_INFO)));
+            Marshal.StructureToPtr(fileInfo, winTrustData.File, false);
```

Comment:
Is this safe? You create an instance of a structure on the managed heap (which may end up behaving like the stack), create a pointer to it, stash that pointer into a native structure, and then pass that structure along to a native API. Does StructureToPtr handle pinning the managed struct or otherwise ensuring that it won't be moved by the GC process later?